source/slang/ir-sccp.cpp

```diff
@@ -0,0 +1,933 @@
+        // on any value whatsoever, then any further update to our
+        // guess can't expand things more, and so there is nothing to do.
+        //
+        if( oldVal.flavor == LatticeVal::Flavor::Any )
+        {
+            return;
+        }
+
+        // Otherwise, we compute a new guess at the value of
+        // the instruction based on the lattice values of the
+        // stuff it depends on.
+        //
+        LatticeVal newVal = interpretOverLattice(inst);
+
+        // If nothing changed about our guess, then there is nothing
+        // further to do, because users of this instruction have
+        // already computed their guess based on its current value.
+        //
+        if(newVal == oldVal)
+        {
+            return;
+        }
+
+        // If the guess did change, then we want to register our
+        // new guess as the lattice value for this instruction.
+        //
+        setLatticeVal(inst, newVal);
+
+        // Next we iterate over all the users of this instruction
+        // and add them to our work list so that we can update
+        // their values based on the new information.
+        //
+        for( auto use = inst->firstUse; use; use = use->nextUse )
+        {
+            ssaWorkList.Add(use->getUser());
+        }
+    }
+
+    // The `apply()` function will run the full algorithm.
+    //
+    void apply()
+    {
+        // We start with the busy-work of setting up our IR builder.
+        //
+        builderStorage.sharedBuilder = &shared->sharedBuilder;
+
+        // We expect the caller to have filtered out functions with
+        // no bodies, so there should always be at least one basic block.
+        //
+        auto firstBlock = code->getFirstBlock();
+        SLANG_ASSERT(firstBlock);
+
+        // The entry block is always going to be executed when the
+        // function gets called, so we will process it right away.
+        //
+        cfgWorkList.Add(firstBlock);
+
+        // The parameters of the first block are our function parameters,
+        // and we want to operate on the assumption that they could have
+        // any value possible, so we will record that in our dictionary.
+        //
+        for( auto pp : firstBlock->getParams() )
+        {
+            setLatticeVal(pp, LatticeVal::getAny());
+        }
+
+        // Now we will iterate until both of our work lists go dry.
+        //
+        while(cfgWorkList.Count() || ssaWorkList.Count())
+        {
+            // Note: there is a design choice to be had here
+            // around whether we do `if if` or `while while`
+            // for these nested checks. The choice can affect
+            // how long things take to converge.
+
+            // We will start by processing any blocks that we
+            // have determined are potentially reachable.
+            //
+            while( cfgWorkList.Count() )
+            {
+                // We pop one block off of the work list.
```

Comment:
It says 'pop' - implying stack like, but it looks as if that's not important. If any element can be taken, removing from the back is the faster op.